### PR TITLE
quasistatic noise-susceptibility

### PR DESCRIPTION
**Pull request by ViktorPerunicic**
pending core bug resolution 

@cdbbentley - hear you'll be reading this NB - let me know if you spot typos ;) !

### CODE REVIEW

File: qctrl/Evaluating quasistatic noise-susceptibility of dynamically corrected gates.ipynb [hunk 1]
```diff
@@ -0,0 +1,546 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Evaluating quasistatic noise-susceptibility of dynamically corrected gates \n",
```
Comment (cdbbentley) [inline qctrl/Evaluating quasistatic noise-susceptibility of dynamically corrected gates.ipynb]: "Evaluating susceptibility of dynamically corrected gates to quasistatic noise" may be clearer?

File: qctrl/Evaluating quasistatic noise-susceptibility of dynamically corrected gates.ipynb [hunk 2]
```diff
@@ -0,0 +1,546 @@
...
+    "\n",
+    "### Calculation of infidelity response of driven controls to simultaneous dephasing and amplitude quasistatic noise\n",
+    "\n",
+    "\n",
+    "The Q-CTRL Python Package enables you to evaluate the quasistatic noise susceptibility of your quantum controls as a function of up to two simulations noise channels.\n",
```
Comment (cdbbentley) [inline qctrl/Evaluating quasistatic noise-susceptibility of dynamically corrected gates.ipynb]: simulations -> simultaneous

File: qctrl/Evaluating quasistatic noise-susceptibility of dynamically corrected gates.ipynb [hunk 3]
```diff
@@ -0,0 +1,546 @@
...
+    "\n",
+    "In this notebook, we present an example of performing a 2D quasistatic noise evaluation for $\\pi$ pulses robust to dephasing and amplitude noise from [Open Controls](https://q-ctrl.com/products/open-controls/) and for pulses constructed using the robust [optimization](https://docs.q-ctrl.com/optimization) feature of the Q-CTRL Python Package. The schemes considered to perform pulses are the following:\n",
+    "    * primitive,\n",
+    "    * BB1\n",
+    "    * CORPSE\n",
+    "    * An optimized pulse robust against dephasing and amplitude noise.\n",
```
Comment (cdbbentley) [inline qctrl/Evaluating quasistatic noise-susceptibility of dynamically corrected gates.ipynb]: We consider the following pulse schemes:
 * Primitive, ... and I would add commas after BB1 and CORPSE for consistency

File: qctrl/Evaluating quasistatic noise-susceptibility of dynamically corrected gates.ipynb [hunk 4]
```diff
@@ -0,0 +1,546 @@
...
+   "source": [
+    "## Define your Quantum System\n",
+    "\n",
+    "In the following examples, we will compare a series of composite $\\pi$ pulses applied to a single qubit under amplitude and dephasing noise. The Hamiltonian of the quantum system is:\n",
+    "\n",
+    "\\begin{align*}\n",
+    "H(t) = &\\frac{1+\\beta_\\Omega(t)}{2}\\left( \\Omega(t) \\sigma_- + \\Omega^*(t) \\sigma_+ \\right) + \\frac{\\Delta(t)}{2}\\sigma_z +  \\frac{\\eta(t)}{2} \\sigma_z\n",
+    "\\end{align*}\n",
+    "\n",
+    "where $\\Omega(t)$ is a time dependent Rabi rate created by a microwave drive, $\\beta_\\Omega(t)$ is a fractional time-dependent amplitude fluctuation process, $\\Delta(t)$ is the time dependent frequency shift,  $\\eta(t)$ is a small stochastic slowly varying dephasing noise process and $\\sigma_k$ are the Pauli matrices.\n",
```
Comment (cdbbentley) [inline qctrl/Evaluating quasistatic noise-susceptibility of dynamically corrected gates.ipynb]: time-dependent Rabi rate... time-dependent frequency shift...slowly-varying dephasing noise process

File: qctrl/Evaluating quasistatic noise-susceptibility of dynamically corrected gates.ipynb [hunk 5]
```diff
@@ -0,0 +1,546 @@
...
+    "\n",
+    "and two noises:\n",
+    "\n",
+    "* A `control noise` associated with the microwave drive control,\n",
+    "* An `additive noise` with `operator` $\\sigma_z$.\n",
+    "\n",
+    "\n",
+    "We first need to parametrize the controls, as described in [Control Formats](https://docs.q-ctrl.com/control-formats). Each pulse is treated as a piecewise-constant function made of a set of `segments`. Each `segment` has a `duration` in time and a `real`/`complex` `value` that is held constant.\n",
+    "\n",
+    "Lastly, we prepare a list of quasistatic noise coefficients `x_coefficients` and `y_coefficients` (corresponding to $\\beta$ and $\\eta$), load them into the system using `QuasiStaticFunction`, and evaluate the gate infidelity as a function of all the combinations of coefficient pairs."
```
Comment (cdbbentley) [inline qctrl/Evaluating quasistatic noise-susceptibility of dynamically corrected gates.ipynb]: \\beta -> \\beta_\\Omega

File: qctrl/Evaluating quasistatic noise-susceptibility of dynamically corrected gates.ipynb [hunk 6]
```diff
@@ -0,0 +1,546 @@
...
+    "sigma_z = np.array([[1., 0.],[0., -1.]], dtype=np.complex)\n",
+    "sigma_x = np.array([[0., 1.],[1., 0.]], dtype=np.complex)\n",
+    "sigma_m = np.array([[0., 0.],[1., 0.]], dtype=np.complex)\n",
+    "\n",
+    "# Define maximum driving rate\n",
+    "omega_max = 2 * np.pi *  1e6 #Hz\n",
+    "total_rotation = np.pi\n",
+    "azimuthal_angle = 0\n",
+    "\n",
+    "# QuasiStaticFunction paramters\n",
```
Comment (cdbbentley) [inline qctrl/Evaluating quasistatic noise-susceptibility of dynamically corrected gates.ipynb]: parameters

File: qctrl/Evaluating quasistatic noise-susceptibility of dynamically corrected gates.ipynb [hunk 7]
```diff
@@ -0,0 +1,546 @@
...
+    "quasi_static_functions = {}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Open Control pulses\n",
+    "\n",
+    "To help you quantify how a control pulse responds to the control noise power spectrum,  the package provides you with `services` to calculate its associated [Filter Function](https://https://docs.q-ctrl.com/filter-functions).\n",
```
Comment (cdbbentley) [inline qctrl/Evaluating quasistatic noise-susceptibility of dynamically corrected gates.ipynb]: 'responds to the noise power spectrum, ...'

File: qctrl/Evaluating quasistatic noise-susceptibility of dynamically corrected gates.ipynb [hunk 8]
```diff
@@ -0,0 +1,546 @@
...
+    "    \n",
+    "    # set up quasistatic function\n",
+    "    quasi_static_function = qctrl.factories.quasi_static_functions.create(\n",
+    "            x_noise=dephasing_noise,\n",
+    "            x_coefficients=dephasing_coefficients,\n",
+    "            y_noise=amplitude_noise,\n",
+    "            y_coefficients=amplitude_coefficients    \n",
+    "    )\n",
+    "\n",
+    "    # calculate quasi static function\n",
```
Comment (cdbbentley) [inline qctrl/Evaluating quasistatic noise-susceptibility of dynamically corrected gates.ipynb]: 'quasistatic' for consistency

File: qctrl/Evaluating quasistatic noise-susceptibility of dynamically corrected gates.ipynb [hunk 9]
```diff
@@ -0,0 +1,546 @@
...
+    "  \n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Optimized pulse\n",
+    "\n",
+    "In this example we make an optimized control pulse for an single qubit which is robust to both the amplitude and dephasing noise. Our aim is to construct control pulses that achieve a X gate, i.e.\n",
```
Comment (cdbbentley) [inline qctrl/Evaluating quasistatic noise-susceptibility of dynamically corrected gates.ipynb]: 'a single qubit', 'an X gate'

File: qctrl/Evaluating quasistatic noise-susceptibility of dynamically corrected gates.ipynb [hunk 10]
```diff
@@ -0,0 +1,546 @@
...
+    "\n",
+    "\\begin{align*}\n",
+    "U_X = \\mathcal{T}\\exp\\left(-i\\int_0^T H(t) dt \\right),\n",
+    "\\end{align*}\n",
+    "\n",
+    "as in the Open Controls pulses considered above. Here, you only need to define a set of constraints and the `robust optimization` service creates pulse segments. Each `drive control`  has an `upper_bound` on its modulus, and it can be optionally constrained to have a `fixed_modulus`. All pulses have the same total `duration` $T$ and the same `segment_count`. \n",
+    "\n",
+    "The optimizer finally needs a `target` operation to achieve. This is described by a `unitary_operator` and `projection_operator`. The projector operator is useful if you want only achieve a unitary in some subspace and are not worried about the relative phase of this unitary.\n"
```
Comment (cdbbentley) [inline qctrl/Evaluating quasistatic noise-susceptibility of dynamically corrected gates.ipynb]: "The projector operator is useful if you only want to achieve...'
Note: @michaelhush and I were talking about the use of we/you but this is a bigger question across all the notebooks.

File: qctrl/Evaluating quasistatic noise-susceptibility of dynamically corrected gates.ipynb [hunk 11]
```diff
@@ -0,0 +1,546 @@
...
+    "                        clock_shift_pulse=robust_optimized_control['Clock'])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Compare\n",
+    "\n",
+    "Here we compare the amplitude and dephasing quasistatic noise response for all the pulses. The optimized pulses outperform the analytic pulses both in robustness and speed. For example, observe that the plot area with infidelity less than $10^{-3}$ is around 3 times larger for the optimized pulse compared to analytic. Also, the optimized pulse simultaneously suppresses both noise types, yet it is approximately 20% faster than eather CORPSE or BB1 for the same maximum Rabi rate."
```
Comment (cdbbentley) [inline qctrl/Evaluating quasistatic noise-susceptibility of dynamically corrected gates.ipynb]: 'is around three times larger for the optimized pulse compared to the analytic pulse' (Or pulses? - is this all the above cases?).
'eather' -> 'either'

File: qctrl/Evaluating quasistatic noise-susceptibility of dynamically corrected gates.ipynb [hunk 12]
```diff
@@ -0,0 +1,546 @@
...
...P3pB2kTa+KtpmaYwTtwISAiCwDfgc4gT8rpf7fAOW+DPwdmK+UWj+EJmpinL62eBren0T9mgl4OuNIKqon/5wtpM+o0poCzbAiPqedgvM2M/qMbdR/WETd6lL2LV9EXFYbOSdvJ2NuGQ53bD65a4Y/+i2kEBARJ3AfcAZQAawTkReUUlsDyqUC3wbWDr2VmlilpymJulWTaVxXgupzkja9itxTtpNc3KAFuZphjTOhj5wlu8hetIuWLQXUrpxE1fNzOfjmNLKX7CTrhF3g1lmyNeEllLeQokXMODDAAmCXUmoPgIg8BVwAbA0odxfwP8APrDQaGAfGju7CTh0z/UewOna0KVZ0M2YaGDu6GTt6FjNdTbAyZv121ydT9dZUGj4qBCBrbjljlu4mcXR/+Pa4iGmY7OTQikQcGDtxNWJ5aNiOniUcMV2s1DHTngT7LgXWMdOZDNZG7txaco6vpXlHJjVvTaLm1eOoWzGFnCW7yTlpJ66kQ8HxzHIhWcllZeVaM9PWhEPPYkfXFUzPYqZZsmKbFZ3MSCGW7xMQWw5MAVDut1wBLPQvICJzgEKl1L9EZEAHRkRuAm4CYt6D1NijpyGJmrem0LC+GHEochfvYdQpu4jL7NT6Fs2IRgRSS+tILa2jozyD6rcmUfPmVGpXTyD35J3kLtmFM/HY+ZHVRA7twFgn2JH6/HFARBzA/wFfNWtIKfUA8ABARkaGFj6MIHqbE6h5cyoN68aBKHIW7WH00h2402I77btGEwmSCpsY/9UPaa9IpeaNqdS8Pp26dyeSe+oOsk7cjiNOTy1p7CEi2oEJgQqg0G95LOAvt08FZgArfAd1NPCCiJyvhbwjH0+Hm9pVU6h/byJ4HWQtKGPUaduJy+jUGhfNMU9ifjPjrv2Ajsp0al6fRvUrM6hbPYHcpVvJmr9HC9g1IeP1erWINwTWARNFpASoBC4DrujfqJRqBnL6l0VkBfAD7byMbLy9TurXlFK3YgqerjgyZpcz+qytxGe3R9s0jSbmSCpopuS692nfm0XVSzM48Pxc6t+dxKgzN5MyfS8x/kCtiSG0iDcElFJ9InIr8BrGa9QPK6W2iMidwHql1At22zZL5mgm2AwmjDUL7GYnoaCZ6DVYGSuB3sIRMM/O/h2NeFgpaN44luqXZ9LTmEza1GrGnLOF1MJ+x8U9oK12jn0kAtlZST
```
Comment (cdbbentley) [inline qctrl/Evaluating quasistatic noise-susceptibility of dynamically corrected gates.ipynb]: \beta_\Omega on the axis for consistency
Do we want to explicitly say Infidelity somewhere?

**Comment by michaelhush** [comment]
I'm going to just state more changes needed, until @ViktorPerunicic you've addressed the requests from Chris.

**Comment by ViktorPerunicic** [comment]
@michaelhush planing to come back to this notebook as soon as PRODUCT-1274 is no longer blocking it... or have they deployed the updated core version?

**Comment by michaelhush** [comment]
Yeah no problem. I was just clearing my PR requests. No it's still in the backlog.

**Comment by ViktorPerunicic** [comment]
@michaelhush.  I incorporated Chris's typos & suggestions (note I have left the "you/we" style as in line with other NB, remind me where we left off from that discussion)

**Pull request merged**